isscrolls.1
.\"
.\" Copyright (c) 2021-26 Matthias Schmidt
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.\"
.Dd February 25, 2026
.Dt ISSCROLLS 1
.Sh NAME
.Nm isscrolls
.Nd Player toolkit for the Ironsworn Family Tabletop RPG
.Sh SYNOPSIS
.Nm isscrolls
.Op Fl bcx
.Sh DESCRIPTION
.Nm
is a toolkit for players of the
.Em Ironsworn
tabletop RPG.
It supports classic
.Em Ironsworn , Delve , Starforged ,
and
.Em Sundered Isles .
It is intended for both solo and co-op play and allows tracking your
character's progress including vows, roll on adventure and exploration moves,
tracks progress for fights, journeys, and expeditions and provides results
from the static oracle tables from the official rulebook.
The options are as follows:
.Bl -tag -width Ds
.It Fl b
Suppress the banner on startup.
.It Fl c
Enable colors and additional characters to beautify output.
Recommended if you don't use a screen reader or a braille terminal.
.It Fl x
Roll a
.Dq cursed die
alongside the other dice.
This is a speciality of the
.Em Sundered Isles
expansion, however, it can be used for the other games as well.
.El
.Sh HOW TO USE
.Nm
basically works like a
.Ux
shell.
Commands can be entered and will show their output.
Certain commands need an argument in order to run properly, other commands
might support optional arguments.
Once a character is loaded, the prompt will change and display the
character's name.
The
.Ic help
command will show an overview of all available commands.
.Pp
.Nm
is linked against
.Xr readline 3 ,
thus any shortcut or character combination that work with a common
.Ux
shell also work for the built-in shell.
.Ss Information for general Gameplay
.Nm
supports all games of the
.Em Ironsworn
family and allows players to freely
use moves or actions from all of them.
Therefore, depending on the game you own you might see statistics or
information that belong to another game, e.g., characters get experience
points in classic
.Em Ironsworn
for fulfilling vows while in
.Em Starforged
they
get experience via the so called
.Em legacy track .
Both ways are supported by
.Nm
so choose the one that fits your game the most.
.Pp
Sometimes
.Nm
cannot take a decision for you and refers you to the official rulebook.
In these cases choose the outcome most suitable for your character and
manipulate the stats with the
.Ic increase ,
.Ic decrease
or
.Ic toogle
commands.
See the section below for more information how the commands work.
.Ss Character Display
.Nm
is best used if you generate a character with the
.Ic create
command and load it afterward with the
.Ic cd
command.
Once you load a character, their current stats will be printed which looks
similar to the following:
.Bd -literal -offset indent
Name: Erin Kalina (Exp: 7/30) Exp spent: 4

Edge: 4 Heart: 3 Iron: 2 Shadow: 2 Wits: 3

Momentum: 10/10 [2] Health: 4/5 Spirit: 5/5 Supply: 3/5

Wounded:        0 Unprepared:   0 Encumbered:   0 Shaken:       0
Corrupted:      0 Tormented:    0 Cursed:       0 Maimed:       1

Bonds: 0.00 (L: 0) Quests: 0.00 (L: 0) Discoveries: 0.00 (L: 0)
.Ed
.Pp
The first line shows the character's name, their experience points, the
amount of experience points spent to acquire new resources or improve
existing ones and their failure track in ticks in case the character has rolled
more than one miss.
.Pp
The second line shows the character's stats and the third one the current
values.
The number in the square brackets after
.Em Momentum
is the reset momentum.
.Pp
Debilities are shown in the next two lines, a 0 means that the character
does not suffer from this particular debility while a 1 means that they
suffer from it.
.Pp
The last line shows the character's progress for bonds, quests and,
discoveries including the progress on the legacy tracks.
.Ss Vows, Expeditions, Journeys, Fights and Delve
As soon as a character starts one of them, the prompt changes and
displays the current undertaking and the current progress.
.Pp
Progress will be tracked automatically according to the rank.
For lower ranks (Troublesome - Formidable),
progress will be shown as absolute numbers, e.g. a 2 represents
two boxes.
For higher ranks (Extreme and Epic), progress will be shown as decimal
numbers and 0.25 represents one tick,
e.g. 0.75 means that the character already made 3 ticks progress.
Since
.Em Ironsworn's
progress bar is always 10 boxes, only the current progress is shown.
.Pp
In case the rulebook gives you more options, you always have the possibility
to manually mark progress with the
.Ic markprogress
command.
In case you have an active vow, its title and description will be shown.
.Ss Notes and Journaling
It is possible to create notes for future reference.
.Pp
It is also able to write entries to a journal file.
.Pp
.Em Autojournalling
can be turned on to automatically journal some commands and their results,
such as the results of random actions.
.Sh AVAILABLE COMMANDS
The following sections shows all commands that can be entered on
.Nm
command prompt.
.Ss General Commands
This subsection shows general commands that are not dice rolls and game moves.
.Bl -tag
.It Ic cd Op name
Load the character
.Op name ,
if it exists.
If
.Op name
is not provided and a character is loaded, the character is saved and unloaded.
.It Ic help
Show an overview of all available commands.
.It Ic ls
List all available characters.
.It Ic quit
Quits
.Nm
and saves all characters, journeys, fights, delves, vows  and the command line
history.
.It Ic save
Saves the current character including an active vow, journey, fight, or delve.
.It Ic startautojournal
Starts autojournalling.
When autojournalling is on,
some commands and their results will be written to the journal file.
.It Ic stopautojournal
Stops autojournalling.
.It Ic journal Cm text
Writes
.Cm text
as an entry to a journal file.
This command can be abbreviated to 'j'.
.Pp
The file will be in the isscrolls directory.
There will be a separate file for each character.
If the current character is deleted and a new character is created with the same
name, a new journal file will be created.
.El
.Ss Dice Rolls
The following commands can be used to roll dice according to the game's
mechanics.
They can be used since
.Nm
does not provide commands for all moves in the game.
.Bl -tag
.It Ic action Cm stat Op bonus
Roll one
.Em action die
and two
.Em challenge dice .
Providing one number as
.Cm stat
is mandatory.
Provide an optional
.Op bonus .
.It Ic challenge
Roll one
.Em challenge die .
.It Ic oracle
Roll one
.Em oracle die .
.It Ic markabond
Mark a bond.
Usually, this is done automatically if you have a strong hit on the
.Ic forgeabond
move.
On a weak hit, consult the Rulebook first and then use this command.
.It Ic markprogress
Mark progress according to the rank.
.Nm
progresses the undertaking in the following order:
.Bl -enum -compact
.It
Fight
.It
Delve
.It
Journey
.El
.It Ic yesorno Cm odds
Roll two
.Em challenge dice
to get an answer to a yes/no question from the oracle.
.Cm odds
has to be a number (1-5) of the following list:
.Bl -enum -compact
.It
Almost certain
.It
Likely
.It
50/50
.It
Unlikely
.It
Small chance
.El
.It Ic burnmomentum
Burn your character's momentum and set it to the reset momentum.
.El
.Ss Character Commands
The following commands can be used to modify your character.
.Nm
supports multiple characters, however, only one can be loaded at a time.
Most commands do nothing or present a warning if no character is loaded.
.Bl -tag
.It Ic create Op name
Create a new character.
If the optional argument
.Op name
is provided, it will be used as the character's name.
Otherwise, the name will be asked interactively.
.It Ic decrease Cm stat
Decrease the character's value
.Cm stat .
You can also decrease the progress of a fight or a journey.
Hereby, the progress is automatically decreased according to the rank.
You can also decrease the progress of or the quest,
discovery and failure track.
.Pp
To give the player a bit more flexibility, this allows to decrease values even
if this is forbidden by the official rulebooks.
.Pp
The following values can be decreased:
.Bl -bullet -compact
.It
Edge
.It
Heart
.It
Iron
.It
Shadow
.It
Wits
.It
Momentum
.It
Health
.It
Spirit
.It
Supply
.It
Exp
.It
Expspent (the experience you spent to acquire new resources or improve
existing ones).
.It
Weapon (hereby
.Em 1
means a simple weapon doing 1 harm and
.Em 2
means a deadly weapon doing 2 harm).
.It
failure (progress on the failure track)
.It
quests (progress on the quests (vows) track)
.It
discoveries (progress on the discoveries track)
.El
.It Ic delete
Delete the current character without warning!
.It Ic increase Cm stat
Opposite of the
.Ic decrease
command described above.
.It Ic print
Print the current character's stats.
.It Ic toggle Cm stat
Toggle a character's stat named
.Cm stat .
If it's set to 0, it will be set to 1 or vice versa.
The following values can be modified.
.Bl -bullet -compact
.It
Wounded
.It
Unprepared
.It
Shaken
.It
Encumbered
.It
Maimed.
Note that this is a permanent bane and cannot be changed once it has been set.
.It
Cursed.
Note that this is a permanent bane and cannot be changed once it has been set.
.It
Corrupted
.It
Tormented
.El
.El
.Ss Vow Management
The following commands let you manage vows.
Additional vow related commands can be seen below in the
Quest Move section.
.Bl -tag
.It Ic vownew
Compared to
.Em swearanironvow ,
this command just saves a new vow and does not roll the die.
It is meant to be used to save the character's initial vows.
.Pp
You will be asked for a title - which is like a short description of the vow -
and a longer description.
The prompt changes and shows the title of the active vow and the current
progress.
.It Ic vowactivate Cm id
Activate the vow
.Cm id ,
with
.Cm id
being a number between 1 and 255.
Every vow has a unique number that can be seen with the
vowshow command.
Activating a vow means that the prompt shows the title and certain moves
will influence it.
.It Ic vowdeactivate
Opposite command which deactivates the current vow.
Certain vow related commands will not work.
.It Ic vowshow
Shows the character's vows including the following details:
.Bl -bullet -compact
.It
ID that can be used with vowactivate and vowdeactivate
.It
Title of every vow
.It
Current progress
.It
The rank
.It
Fulfillment status.
0 means unfulfilled, 1 means fulfilled.
.El
.El
.Ss Notes Management
The following commands let you manage 'sticky' notes.
.Bl -tag
.It Ic notenew
Creates a 'sticky' note.
Every note has a unique number that can be seen with the
noteshow command.
Each note must have a title and a description.
.It Ic noteedit Cm id
Edit an existing note.
Modify the title and description.
.It Ic noteshow
Shows the character's notes including the following details:
.Bl -bullet -compact
.It
ID that can be used with noteedit and notedelete.
.It
Title of every note
.It
Description of every note
.El
.It Ic notedelete Cm id
Deletes an existing note.
.El
.Ss Adventure and Exploration Moves
Adventure Moves are used as your character travels the Ironlands, investigates
situations and deals with threats.
Exploration moves are the same concept, however, the new name was introduced
with
.Em Ironsworn Starforged .
.Bl -tag
.It Ic facedanger Cm stat Op bonus
Roll a
.Em Face Danger
move using the character's stat named
.Cm stat .
This move can be done using the following stats: edge, heart, iron, shadow, and
wits.
An additional
.Op bonus
can be provided.
.It Ic secureanadvantage Cm stat Op bonus
Roll a
.Em Secure an Advantage
move using the character's stat named
.Cm stat .
This move can be done using the following stats: edge, heart, iron, shadow, and
wits.
An additional
.Op bonus
can be provided.
.It Ic gatherinformation Op bonus
Roll a
.Em Gather Information
move.
Provide an optional
.Op bonus .
.It Ic heal Op bonus
Roll a
.Em Heal
move.
This move can be done using the following stats: heart, iron, and wits.
An additional
.Op bonus
can be provided.
.It Ic hearten
Roll a
.Em Hearten
move.
.It Ic resupply Op bonus
Roll a
.Em Resupply
move.
This move can be done using the following stats: heart, iron, shadow, and wits.
An additional
.Op bonus
can be provided.
.It Ic makecamp Op bonus
Roll a
.Em Make Camp
move.
.It Ic exploreawaypoint
Roll an
.Em Explore a Waypoint
move
Note that a strong hit with a match provides you with a different result..
.It Ic setacourse Op bonus
Roll an
.Em Set a course
move.
Provide an optional
.Op bonus .
.It Ic undertakeanexpedition Cm stat Op bonus
Roll an
.Em Undertake an Expedition
move using the character's stat named
.Cm stat .
This move can be done using the following stats: edge, shadow, and wits.
Provide an optional
.Op bonus .
.Pp
In case this is the first move of a new expedition,
.Nm
will ask for a rank and save it for the character.
.It Ic undertakeajourney Op bonus
Roll an
.Em Undertake a Journey
move.
Provide an optional
.Op bonus .
.Pp
In case this is the first move of a new journey,
.Nm
will ask for a rank and save it for the character.
.It Ic finishanexpedition Op bonus
Roll a
.Em Finish an Expedition
move.
In case of a
.Em miss
, the character can choose to abort the expedition or continue with it.
An additional
.Op bonus
can be provided.
.It Ic reachyourdestination Op bonus
Roll a
.Em Reach Your Destination
move.
In case of a
.Em miss
, the character can choose to abort the journey or continue with it.
An additional
.Op bonus
can be provided.
.El
.Ss Relationship Moves
The following moves are made as the character interacts with others in the wild,
fight duels, form bonds, supports their allies, and determines the ultimate
fate.
.Bl -tag
.It Ic compel Cm stat Op bonus
Roll a
.Em Compel
move using the character's stat named
.Cm stat .
This move can be done using the following stats: heart, iron, and shadow.
An additional
.Op bonus
can be provided.
.It Ic sojourn Op bonus
Roll a
.Em Sojourn
move.
Provide an optional
.Op bonus .
.It Ic drawthecircle Op bonus
Roll a
.Em Draw The Circle
move.
Provide an optional
.Op bonus .
.It Ic makeaconnection Op bonus
Roll a
.Em Make a Connection
move.
An additional
.Op bonus
can be provided.
Note that
.Nm
does not ask you for a rank and does not track connections, this is up to
the player.
.It Ic forgeabond Op bonus
Roll a
.Em Forge a Bond
move.
An additional
.Op bonus
can be provided.
.It Ic testyourbond Op bonus
Roll a
.Em Test Your Bond
move.
An additional
.Op bonus
can be provided.
.It Ic testyourrelationship Op bonus
Roll a
.Em Test Your Relationship
move.
An additional
.Op bonus
can be provided.
.It Ic writeyourepilogue
Roll a
.Em Write your epilogue
move.
.El
.Ss Combat Moves
When there are no other options, when the sword flips free of its sheath, when
the arrow is nocked, when the shield is brought to bear, these moves can
be made.
.Bl -tag
.It Ic enterthefray Cm stat Op bonus
Roll an
.Em Enter the Fray
move using the character's stat named
.Cm stat .
This move can be done using the following stats: heart, wits, and shadow.
.Pp
In case this is the first move in a fight,
.Nm
will ask for a rank and save it for the fight.
Progress per harm will be tracked automatically according to the rank.
For lower ranks (Troublesome - Formidable), progress will be shown as absolute
numbers, e.g. 2/10.
For higher ranks (Extreme and Epic) progress will be shown as decimal
numbers and 0.25 represents one tick,
e.g. 0.75/10 means that the character already made 3 ticks progress.
An additional
.Op bonus
can be provided.
.It Ic gainground Cm stat Op bonus
Roll an
.Em Gain Ground
move using the character's stat named
.Cm stat .
This move can be done using all stats.
An additional
.Op bonus
can be provided.
.It Ic reactunderfire Cm stat Op bonus
Roll an
.Em React under Fire
move using the character's stat named
.Cm stat .
This move can be done using all stats.
An additional
.Op bonus
can be provided.
.It Ic endthefight Op bonus
Roll an
.Em End the Fight
move.
An additional
.Op bonus
can be provided.
.Nm
checks automatically that your last move was a strong hit.
.It Ic strike Cm stat Op bonus
Roll a
.Em Strike
move using the character's stat named
.Cm stat .
This move can be done using the following stats: iron, and edge.
An additional
.Op bonus
can be provided.
.It Ic clash Cm stat Op bonus
Roll a
.Em Clash
move using the character's stat named
.Cm stat .
This move can be done using the following stats: iron, and edge.
An additional
.Op bonus
can be provided.
.It Ic battle Cm stat Op bonus
Roll a
.Em Battle
move using the character's stat named
.Cm stat .
This move can be done using the following stats: edge, heart, iron, shadow, and
wits.
An additional
.Op bonus
can be provided.
.It Ic getinitiative
Give the character initiative/let them take control.
This can be useful when a player rolls a weak hit on
.Em Enter the Fray .
.El
.Ss Quest Moves
The commands represent important moves characters make during their quest.
.Bl -tag
.It Ic fulfillyourvow
Roll a
.Em Fulfill your Vow
move.
The active vow will be set as fulfilled so that the player cannot activate
it again.
Your character receives experience points equal to the rank of the vow.
.It Ic forsakeyourvow
Roll a
.Em Forskace your Vow
move.
The active vow will be forsaken and removed from the list of vows.
Your character will endure stress equal to the rank of the quest.
.It Ic reachamilestone
Roll a
.Em Reach a Milestone vow
move.
The progress of the active vow is advanced equal to the rank.
.It Ic swearanironvow Op bonus
Roll a
.Em Swear an Iron Vow
move.
You will be asked for a title - which is like a short description of the vow -
and a longer description.
The prompt changes and shows the title of the active vow and the current
progress.
Provide an optional
.Op bonus .
.El
.Ss Suffer Moves
These moves are made as a result of a perilous event or bad outcome on other
moves.
They represent what happens to the character, and how they hold up against
the trauma.
.Bl -tag
.It Ic endureharm Op value
Roll an
.Em Endure Harm
move.
In case the character is in a fight, the amount of harm to suffer is
automatically derived from the foe's rank.
If the character is not in a fight, the amount of harm to suffer can be
provided with the optional argument
.Op value .
Since there are multiple options on what can happen on a
.Dq miss ,
the decision is up to the player and not implemented in
.Nm .
.It Ic facedeath Op bonus
Roll a
.Em Face Death
move.
In case the character dies, it is marked as such and
.Dq Deceased
is printed in the character's statistics.
An additional
.Op bonus
can be provided.
.It Ic facedesolation Op bonus
Roll a
.Em Face Desolation
move.
An additional
.Op bonus
can be provided.
.It Ic endurestress Cm value
Roll an
.Em Endure Stress
move.
Provide a
.Cm value
that will be removed from your spirit.
Since there are multiple reasons why the character can suffer stress, the
amount of
.Em Spirit
to suffer has to be provided and is not implemented within
.Nm .
.It Ic sacrificeresources Cm value
Roll an
.Em Sacrifice Resources
move.
Provide a
.Cm value
that will be removed from your supply.
.El
.Ss Delve Moves
These moves are part of the
.Em Ironsworn Delve
supplement.
They will help you delve into sites and hopefully locate your objective.
.Bl -tag
.It Ic learnfromyourfailures
Roll a
.Em Learn from your Failures
move.
This is a progress move where the result is compared against the failure
track.
It can only be used if the character has more than 6 boxes marked in the
failure track.
The latter is automatically tracked by
.Nm
and reset after this move.
Although this is a character move, it is part of the
.Em Delve
supplement.
.It Ic discoverasite
Roll a
.Em Discover a Site
move.
This is the first move towards a delve into a site.
.Nm
will ask for the site's rank.
.It Ic delvethedepths Cm stat Op bonus
Roll a
.Em Delve the Depths
move using the character's stat named
.Cm stat .
This move can be done using edge, shadow, and wits.
An additional
.Op bonus
can be provided.
.It Ic checkyourgear Op bonus
Roll a
.Em Check your Gear
move.
An additional
.Op bonus
can be provided.
.It Ic locateyourobjective
Roll a
.Em Locate your Objective
move.
.It Ic escapethedepths Cm stats Op bonus
Roll an
.Em Escape the Depths
This move can be done using edge, heart, iron, shadow, and wits.
An additional
.Op bonus
can be provided.
.El
.Ss Oracle Moves
The following commands provide results from the various oracle tables in both
the
.Em Ironsworn
and
.Em Ironsworn Delve
Rulebook.
.Bl -tag
.It Ic generatenpc
Generate a random NPC with a role, a goal and their disposition.
.It Ic actionoracle
Show a random action.
.It Ic combataction
Show a random combat action.
.It Ic coastalwaterlocation
Show a random coastal waters location.
.It Ic elfname
Show a random elf name.
.It Ic findanopportunity
Show a random opportunity.
.It Ic giantname
Show a random giant name.
.It Ic ironlandername
Show a random Ironlander name.
.It Ic location
Show a random location.
.It Ic locationdescription
Show a random description for a location.
.It Ic moonoracle
Roll random on the
.Em Sundered Isles
moons, Wraith and Cinder.
.It Ic mysticbackslash
Show a random mystic backslash.
.It Ic paytheprice
Show a random
.Dq Pay the price
result.
.It Ic plottwist
Show a random plot twist.
.It Ic rank
Show a random rank.
.It Ic region
Show a random Ironlands region.
.It Ic revealadanger
Show a random danger region.
.It Ic settlementtrouble
Show a random settlement trouble.
.It Ic theme
Show a random theme.
.It Ic varou
Show a random Varou name.
.El
.Sh ENVIRONMENT
.Nm
makes use of the following environment variables.
.Bl -tag -width XDG_CONFIG_HOME
.It Ev HOME
If the
.Ev XDG_CONFIG_HOME
variable is not set,
.Nm
stores its history and other data in the
.Pa .config/isscrolls
subdirectory in the user's home directory.
.It Ev XDG_CONFIG_HOME
In case this variable is set,
.Nm
stores its history and other data in the
.Pa isscrolls
subdirectory below the path set by this environment variable.
.El
.Sh FILES
.Bl -tag -width Ds -compact
.It Pa /usr/local/share/isscrolls
Contains shared files such as the JSON files for the oracle tables.
.El
.Sh EXIT STATUS
.Nm
normally exists with 0 or with 1 if an error occurred.
.Sh SEE ALSO
.Xr readline 3
.Sh STANDARDS
.Rs
.%A Shawn Tomkin
.%B Ironsworn. A tabletop RPG of perilous Quests
.%D 2018
.%U https://www.ironswornrpg.com/
.Re
.Pp
.Rs
.%A Shawn Tomkin
.%B Ironsworn Delve. Perilous Expeditions for the Ironsworn RPG
.%D 2020
.%U https://www.ironswornrpg.com/
.Re
.Pp
.Rs
.%A Shawn Tomkin
.%B Ironsworn Starforged
.%D 2022
.%U https://www.ironswornrpg.com/
.Re
.Sh AUTHORS
.Nm
was written by
.An Matthias Schmidt Aq Mt xhr@giessen.ccc.de .


isscrolls.c
/*
 * Copyright (c) 2021 Matthias Schmidt <xhr@giessen.ccc.de>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/stat.h>
#include <sys/types.h>

#include <errno.h>
#include <limits.h>
#include <signal.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#include <readline/readline.h>
#include <readline/history.h>

#include "isscrolls.h"

static char prompt[MAX_PROMPT_LEN];
static char isscrolls_dir[_POSIX_PATH_MAX];

static int debug = 0;
static int color = 0;
static int cursed = 0;
static int banner = 1;
static int output = 1;

static volatile sig_atomic_t sflag = 0;

FILE *journal_file = NULL;
int journal_this = 0;

static void
signal_handler(int signal)
{
	switch (signal) {
		case SIGINT:
		case SIGTERM:
			sflag = 1;
			break;
	}
}

void
show_banner(__attribute__((unused)) char *unused)
{
	const int BANNER = GREEN | NO_JOURNAL;
	pm(BANNER, "\n██▓  ██████   ██████  ▄████▄   ██▀███   ▒█████   ██▓     ██▓      ██████\n");
	pm(BANNER, "▓██▒▒██    ▒ ▒██    ▒ ▒██▀ ▀█  ▓██ ▒ ██▒▒██▒  ██▒▓██▒    ▓██▒    ▒██    ▒\n");
	pm(BANNER, "▒██▒░ ▓██▄   ░ ▓██▄   ▒▓█    ▄ ▓██ ░▄█ ▒▒██░  ██▒▒██░    ▒██░    ░ ▓██▄\n");
	pm(BANNER, "░██░  ▒   ██▒  ▒   ██▒▒▓▓▄ ▄██▒▒██▀▀█▄  ▒██   ██░▒██░    ▒██░      ▒   ██▒\n");
	pm(BANNER, "░██░▒██████▒▒▒██████▒▒▒ ▓███▀ ░░██▓ ▒██▒░ ████▓▒░░██████▒░██████▒▒██████▒▒\n");
	pm(BANNER, "░▓  ▒ ▒▓▒ ▒ ░▒ ▒▓▒ ▒ ░░ ░▒ ▒  ░░ ▒▓ ░▒▓░░ ▒░▒░▒░ ░ ▒░▓  ░░ ▒░▓  ░▒ ▒▓▒ ▒ ░\n");
	pm(BANNER, " ▒ ░░ ░▒  ░ ░░ ░▒  ░ ░  ░  ▒     ░▒ ░ ▒░  ░ ▒ ▒░ ░ ░ ▒  ░░ ░ ▒  ░░ ░▒  ░ ░\n");
	pm(BANNER, " ▒ ░░  ░  ░  ░  ░  ░  ░          ░░   ░ ░ ░ ░ ▒    ░ ░     ░ ░   ░  ░  ░\n");
	pm(BANNER, " ░        ░        ░  ░ ░         ░         ░ ░      ░  ░    ░  ░      ░\n");
	pm(BANNER, "                      ░\n");
	printf("                                                            Version %s\n\n", VERSION);

	printf("\tPlayer toolkit for the Ironsworn RPG family\n");
	printf("\tBy Matthias Schmidt - https://infosec.exchange/@xhr\n\n");
	printf("Enter 'help' for available commands\n\n");
}

int
main(int argc, char **argv)
{
	char *line, *res;
	int ch;

	/*
	 * Seed the PRNG with the current time of the day.  This is not fine
	 * for a security critical application, for rolling dice it is
	 */
	srandom(time(NULL) ^ getpid());

	while ((ch = getopt(argc, argv, "cdbx")) != -1) {
		switch (ch) {
		case 'b':
			banner = 0;
			break;
		case 'c':
			color = 1;
			break;
		case 'd':
			debug = 1;
			break;
		case 'x':
			cursed = 1;
			break;
		}
	}

	argc -= optind;
	argv += optind;

	setup_base_dir();

	initialize_readline(isscrolls_dir);

	if (banner)
		show_banner(NULL);

	if (signal(SIGINT, signal_handler) == SIG_ERR)
		log_errx(1, "signal");
	if (signal(SIGTERM, signal_handler) == SIG_ERR)
		log_errx(1, "signal");

	sandbox(isscrolls_dir);

	if (load_characters_list() == -1)
		set_prompt("> ");

	while (!sflag) {
		line = readline(prompt);
		if (line == NULL)
			continue;
		res = stripwhite(line);

		if (*res) {
			add_history(res);
			execute_command(res);
		}

		free(line);
	}

	initiate_shutdown(0);

	return 0;
}

void
set_prompt(const char *p)
{
	if (p == NULL || strlen(p) == 0)
		return;

	if (color)
		snprintf(prompt, sizeof(prompt), "%s%s%s", ANSI_COLOR_BOLD, p, ANSI_COLOR_RESET);
	else
		snprintf(prompt, sizeof(prompt), "%s", p);
}

#ifdef __OpenBSD__
void
sandbox(const char *dir)
{
	if (unveil(PATH_SHARE_DIR, "r") == -1)
		log_errx(1, "unveil");
	if (unveil(dir, "rwc") == -1)
		log_errx(1, "unveil");
	if (unveil(NULL, NULL) == -1)
		log_errx(1, "unveil");

	if (pledge("stdio rpath wpath cpath tty", NULL) == -1)
		log_errx(1, "pledge");
}
#else
void sandbox(__attribute__((unused)) const char *dir)
{
}
#endif /* __OpenBSD__ */

void
cmd_quit(__attribute__((unused)) char *unused)
{
	initiate_shutdown(0);
}

void
initiate_shutdown(int exit_code)
{
	char hist_path[_POSIX_PATH_MAX];
	int ret;

	save_current_character();

	ret = snprintf(hist_path, sizeof(hist_path), "%s/history", isscrolls_dir);
	if (ret < 0 || (size_t)ret >= sizeof(hist_path)) {
		printf("Path truncation happened.  Buffer too short to fit %s\n", hist_path);
	}

	log_debug("Writing history to %s\n", hist_path);
	write_history(hist_path);

	close_journal_file();

	exit(exit_code);
}

void
setup_base_dir(void)
{
	struct stat sb;
	char *home, *xdg_home;
	int ret;

	if ((xdg_home = getenv("XDG_CONFIG_HOME")) != NULL) {
		ret = snprintf(isscrolls_dir, sizeof(isscrolls_dir), "%s/isscrolls", xdg_home);
		if (ret < 0 || (size_t)ret >= sizeof(isscrolls_dir)) {
			log_errx(1, "Path truncation happened.  Buffer too short to fit %s\n", isscrolls_dir);
		}
	} else if ((home = getenv("HOME")) != NULL) {
		ret = snprintf(isscrolls_dir, sizeof(isscrolls_dir), "%s/.config/isscrolls", home);
		if (ret < 0 || (size_t)ret >= sizeof(isscrolls_dir)) {
			log_errx(1, "Path truncation happened.  Buffer too short to fit %s\n", isscrolls_dir);
		}
	} else {
		log_errx(1, "Neither $XDG_CONFIG_HOME nor $HOME is set!\n");
	}

	if (stat(isscrolls_dir, &sb) == 0 && S_ISDIR(sb.st_mode)) {
		log_debug("%s already exists\n", isscrolls_dir);
	} else {
		log_debug("%s does not exists.  Attempt to create it\n", isscrolls_dir);
		if (mkdir(isscrolls_dir, 0755) == -1) {
			log_errx(1, "Cannot create %s directory\n", isscrolls_dir);
		}
	}
}

void
log_debug(const char *fmt, ...)
{
	va_list ap;

	if (debug == 0)
		return;

	va_start(ap, fmt);
	fprintf(stdout, "[*] ");
	vfprintf(stdout, fmt, ap);
	va_end(ap);
}

void
log_errx(int prio, const char *fmt, ...)
{
	va_list ap;

	va_start(ap, fmt);
	vfprintf(stderr, fmt, ap);
	va_end(ap);

	initiate_shutdown(prio);
}

void
pm(int what, const char *fmt, ...)
{
	va_list ap;

	if (color && (what & NO_CONSOLE) == 0) {
		switch (what & COLORS) {
		case RED:
			printf("%s", ANSI_COLOR_RED);
			break;
		case YELLOW:
			printf("%s", ANSI_COLOR_YELLOW);
			break;
		case GREEN:
			printf("%s", ANSI_COLOR_GREEN);
			break;
		case BLUE:
			printf("%s", ANSI_COLOR_CYAN);
			break;
		default:
			break;
		}
	}
	if ((what & NO_CONSOLE) == 0) {
		va_start(ap, fmt);
		vprintf(fmt, ap);
		va_end(ap);
	}

	if (color && (what & NO_CONSOLE) == 0) {
		printf("%s", ANSI_COLOR_RESET);
	}

	if ((what & NO_JOURNAL) == 0 && journal_this != 0) {
		va_start(ap, fmt);
		print_to_journal_v(fmt, &ap);
		va_end(ap);
	}
}

const char*
get_isscrolls_dir(void)
{
	return isscrolls_dir;
}

void
toggle_output(void)
{
	output = !output;
}

int
get_output(void)
{
	return output;
}

int
get_color(void)
{
	return color;
}

int
get_cursed(void)
{
	return cursed;
}

void
print_to_journal(const char *format, ...)
{
	va_list args;
	va_start(args, format);
	print_to_journal_v(format, &args);
	va_end(args);
}

void
print_to_journal_v(const char *format, va_list *args)
{

	if (journaling() == 0)
		return;

	if (journal_file == NULL) {
		log_errx(1, "attempting to write to journal but journal not open");
		return;
	}
	vfprintf(journal_file, format, *args);
}

int
journaling(void) {
	struct character *curchar = get_current_character();
	return curchar != NULL && curchar->journaling != 0;
}

void
start_journal_entry(void)
{
	char path[_POSIX_PATH_MAX];
	time_t t;
	struct tm *tm_ptr;

	if (journaling() == 0)
		return;

	if (journal_file == NULL) {
		journal_file_name(path);
		journal_file = fopen(path, "a");
		if (journal_file == NULL) {
			printf("Could not open journal file (%s): %s\n", path, strerror(errno));
			return;
		}
	}

	t = time(NULL);
	tm_ptr = localtime(&t);
	if (tm_ptr == NULL) {
		log_errx(1, "localtime returned null");
		return;
	}
	fprintf(journal_file, "[%d-%02d-%02d %02d:%02d:%02d] ", tm_ptr->tm_year + 1900, tm_ptr->tm_mon + 1, tm_ptr->tm_mday, tm_ptr->tm_hour, tm_ptr->tm_min, tm_ptr->tm_sec);
}

void
close_journal_file(void)
{
	if (journal_file != NULL) {
		fclose(journal_file);
		journal_file = NULL;
	}
}

